apis/compute/v1beta1/firewall_types.go

```diff
@@ -0,0 +1,265 @@
+	// specifies a protocol and port-range tuple that describes a denied
+	// connection.
+	// +optional
+	Denied []*FirewallDenied `json:"denied,omitempty"`
+
+	// Direction: Direction of traffic to which this firewall applies,
+	// either `INGRESS` or `EGRESS`. The default is `INGRESS`. For `INGRESS`
+	// traffic, you cannot specify the destinationRanges field, and for
+	// `EGRESS` traffic, you cannot specify the sourceRanges or sourceTags
+	// fields.
+	//
+	// Possible values:
+	//   "EGRESS"
+	//   "INGRESS"
+	// +optional
+	Direction *string `json:"direction,omitempty"`
+
+	// Disabled: Denotes whether the firewall rule is disabled. When set to
+	// true, the firewall rule is not enforced and the network behaves as if
+	// it did not exist. If this is unspecified, the firewall rule will be
+	// enabled.
+	// +optional
+	Disabled *bool `json:"disabled,omitempty"`
+
+	// LogConfig: This field denotes the logging options for a particular
+	// firewall rule. If logging is enabled, logs will be exported to
+	// Stackdriver.
+	// +optional
+	LogConfig *FirewallLogConfig `json:"logConfig,omitempty"`
+}
+
+// FirewallAllowed represents the ALLOW rule by the firewall
+type FirewallAllowed struct {
+	// IPProtocol: The IP protocol to which this rule applies. The protocol
+	// type is required when creating a firewall rule. This value can either
+	// be one of the following well known protocol strings (tcp, udp, icmp,
+	// esp, ah, ipip, sctp) or the IP protocol number.
+	IPProtocol string `json:"IPProtocol,omitempty"`
+
+	// Ports: An optional list of ports to which this rule applies. This
+	// field is only applicable for the UDP or TCP protocol. Each entry must
+	// be either an integer or a range. If not specified, this rule applies
+	// to connections through any port.
+	//
+	// Example inputs include: ["22"], ["80","443"], and ["12345-12349"].
+	Ports []string `json:"ports,omitempty"`
+}
+
+// FirewallDenied represents the DENY rule by the firewall
+type FirewallDenied struct {
+	// IPProtocol: The IP protocol to which this rule applies. The protocol
+	// type is required when creating a firewall rule. This value can either
+	// be one of the following well known protocol strings (tcp, udp, icmp,
+	// esp, ah, ipip, sctp) or the IP protocol number.
+	IPProtocol string `json:"IPProtocol,omitempty"`
+
+	// Ports: An optional list of ports to which this rule applies. This
+	// field is only applicable for the UDP or TCP protocol. Each entry must
+	// be either an integer or a range. If not specified, this rule applies
+	// to connections through any port.
+	//
+	// Example inputs include: ["22"], ["80","443"], and ["12345-12349"].
+	Ports []string `json:"ports,omitempty"`
+}
+
+// A FirewallLogConfig represents the available logging options for firewall.
+type FirewallLogConfig struct {
+	// Enable: This field denotes whether to enable logging for a particular
+	// firewall rule.
+	Enable bool `json:"enable,omitempty"`
```

Comment:
If this is optional it should be `*bool` and marked with `+ optional`. However, sometimes when there is a single field in an embedded struct I will opt for making it required since the embedded struct really shouldn't be provided if the only field it supports is not set.

Reply:
Right, I agree with that. I have made `FirewallLogConfig` an optional object but made `Enable` as required and removed `omitempty` for it.